File: populateNextRightPointer.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class NodeW:
    def __init__(self, node, level):
        self.node = node
        self.level = level

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if root == None:
            return
        queue = []
        queue.append(NodeW(root,0))
        
        while len(queue) != 0:
            node = queue.pop(0)
            if len(queue) != 0 and queue[0].level == node.level:
                node.node.next = queue[0].node
            else:
                node.node.next = None
            if node.node.left:
                queue.append(NodeW(node.node.left, node.level + 1))
            if node.node.right:
                queue.append(NodeW(node.node.right, node.level + 1))

File: PopulateNextRightPointer.java
/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */
public class Solution {
    
    public TreeLinkNode helper(TreeLinkNode node) {
        TreeLinkNode prev = null;
        TreeLinkNode firstChild = null;
        TreeLinkNode nextChild = null;
        
        if (node == null) {
            return node;
        }
        
        while (node != null) {
            
            while (node != null && node.left == null && node.right == null) {
                node = node.next;
            }
            
            if (node != null && node.left != null && node.right  != null) {
                node.left.next = node.right;
                nextChild = node.right;
                if (prev == null) {
                    firstChild = node.left;
                } else {
                    prev.next = node.left;
                }
            } else {
                nextChild = (node == null) ? null : (node.left == null) ? node.right : node.left;
                if (prev == null) {
                    firstChild = nextChild;
                } else {
                    prev.next = nextChild;
                }
            }
            node = (node == null) ? null : node.next;
            prev = nextChild;
        }
        
        return firstChild;
        
    }
    
    public void connect(TreeLinkNode root) {
        TreeLinkNode head = root;
        TreeLinkNode nextHead = null;
        nextHead = helper(root);
        while (nextHead != null) {
            nextHead = helper(nextHead);
        }
    }
}
